### PR TITLE
Serialize once per format to increase perf in broadcasting scenarios

### PR DESCRIPTION
**Pull request by BrennanConroy**
Also found an issue where we registered KeepAliveTick too soon and fixed it.

### CODE REVIEW

**Comment by BrennanConroy** [comment]
Before:
127k - Json
145k - MsgPack

After:
170k - Json
425k - MsgPack

File: src/Microsoft.AspNetCore.SignalR.Common/Internal/HubProtocolReaderWriter.cs [hunk 1]
```diff
@@ -50,5 +50,23 @@ public byte[] WriteMessage(HubMessage hubMessage)
...
+        public override bool Equals(object obj)
+        {
+            var readerWriter = obj as HubProtocolReaderWriter;
+            if (readerWriter == null)
+            {
+                return false;
+            }
+
+            // Note: ReferenceEquals on HubProtocol works for our implementation of IHubProtocolResolver because we use Singletons from DI
+            // However if someone replaces the implementation and returns a new ProtocolResolver for every connection they wont get the perf benefits
```
Comment (anurse) [inline src/Microsoft.AspNetCore.SignalR.Common/Internal/HubProtocolReaderWriter.cs]: TBH, still nervous about this. It's not even that they won't get the perf benefits, they'll get significantly more memory usage (and `O(n)` memory growth where `n` is the number of connections).

File: src/Microsoft.AspNetCore.SignalR.Common/Internal/HubProtocolReaderWriter.cs [hunk 2]
```diff
@@ -50,5 +50,23 @@ public byte[] WriteMessage(HubMessage hubMessage)
...
+            return ReferenceEquals(_dataEncoder, readerWriter._dataEncoder) && ReferenceEquals(_hubProtocol, readerWriter._hubProtocol);
+        }
+
+        public override int GetHashCode()
+        {
+            return base.GetHashCode();
```
Comment (anurse) [inline src/Microsoft.AspNetCore.SignalR.Common/Internal/HubProtocolReaderWriter.cs]: Comment as to why this is OK (we never want to put this in a hash-table)

File: src/Microsoft.AspNetCore.SignalR.Common/Internal/HubProtocolReaderWriter.cs (hunk 1 above)
Comment (anurse) [reply src/Microsoft.AspNetCore.SignalR.Common/Internal/HubProtocolReaderWriter.cs]: Also, the client is in control of how much this memory grows.

File: src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/HubMessage.cs [hunk 3]
```diff
@@ -1,12 +1,45 @@
...
+            for (var i = 0; i < _serializedMessages.Count; i++)
+            {
+                if (_serializedMessages[i].ProtocolReaderWriter.Equals(protocolReaderWriter))
+                {
+                    return _serializedMessages[i].Message;
+                }
+            }
+
+            var bytes = protocolReaderWriter.WriteMessage(this);
+            _serializedMessages.Add(new SerializedMessage(protocolReaderWriter, bytes));
```
Comment (anurse) [inline src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/HubMessage.cs]: If we capped the cache size, that would relieve some of my concerns about client-controlled linear memory growth from the cache.

File: src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/HubMessage.cs [hunk 4]
```diff
@@ -1,12 +1,45 @@
...
     public abstract class HubMessage
     {
         protected HubMessage()
         {
         }
+
+        // Initialize with capacity 4 for the 2 built in protocols and 2 data encoders
+        private readonly List<SerializedMessage> _serializedMessages = new List<SerializedMessage>(4);
+
+        public byte[] WriteMessage(HubProtocolReaderWriter protocolReaderWriter)
```
Comment (anurse) [inline src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/HubMessage.cs]: Still convinced it's going to bite us that this is on the `HubMessage` type itself.

File: src/Microsoft.AspNetCore.SignalR.Core/HubConnectionContext.cs [hunk 5]
```diff
@@ -84,17 +80,41 @@ public HubConnectionContext(ConnectionContext connectionContext, TimeSpan keepAl
...
+            // Don't wait for the lock, if it returns false that means someone wrote to the connection
+            // and we don't need to send a ping anymore
+            if (!await _writeLock.WaitAsync(0))
+            {
+                return;
+            }
+
+            try
+            {
+                _connectionContext.Transport.Output.Write(_cachedPingMessage);
```
Comment (anurse) [inline src/Microsoft.AspNetCore.SignalR.Core/HubConnectionContext.cs]: Assert `_cachedPingMessage != null`. It depends upon the caller initializing properly so we should assert it.

File: src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/HubMessage.cs (hunk 4 above)
Comment (anurse) [reply src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/HubMessage.cs]: Having this on here is going to mess with the singleton `PingMessage.Instance`... I know you special case that elsewhere but it still makes me nervous dangit!

File: src/Microsoft.AspNetCore.SignalR.Core/HubConnectionContext.cs [hunk 6]
```diff
@@ -154,11 +174,18 @@ public virtual void Abort()
...
                                     ProtocolReaderWriter = new HubProtocolReaderWriter(protocol, dataEncoder);
+                                    _cachedPingMessage = ProtocolReaderWriter.WriteMessage(PingMessage.Instance);
 
                                     Log.UsingHubProtocol(_logger, protocol.Name);
 
                                     UserIdentifier = userIdProvider.GetUserId(this);
 
+                                    if (Features.Get<IConnectionInherentKeepAliveFeature>() == null)
+                                    {
+                                        // Only register KeepAlive after protocol negotiated otherwise KeepAliveTick could try to write without having a ProtocolReaderWriter
```
Comment (anurse) [inline src/Microsoft.AspNetCore.SignalR.Core/HubConnectionContext.cs]: Good comment

File: src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/HubMessage.cs (hunk 3 above)
Comment (davidfowl) [reply src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/HubMessage.cs]: I don’t really understand the concern. Are you worrried about a misconfigured application? We can cap it at 10 if that makes you feel better.

File: src/Microsoft.AspNetCore.SignalR.Common/Internal/HubProtocolReaderWriter.cs (hunk 1 above)
Comment (davidfowl) [reply src/Microsoft.AspNetCore.SignalR.Common/Internal/HubProtocolReaderWriter.cs]: How is it client controlled? The server is the one that configures the amount of protocols supported? Are you taking about in theory you can return a different instance of a protocol per connection? Again how is that a real concern?

**Comment by BrennanConroy** [comment]
Changed the backpressure on the pipes and JSON jumped to ~250k now. We'll have to talk about actual default values, not for this PR though.

**Comment by BrennanConroy** [comment]
:up: 📅

**Pull request merged**